test/cql-pytest/test_validation.py

```diff
@@ -0,0 +1,368 @@
+#    version ("surrogateescape") of bad UTF-8, and monkey-patch the driver to
+#    unescape it when converting it back to bytes.
+# Note that currently, Scylla's UTF-8 parser is stricter than Cassandra's
+# (see comment above listing the relevant cases), so this test, as all tests
+# using the bad_utf8 array, will fail on Cassandra.
+def test_validation_utf8_bound_column(cql, table1):
+    import cassandra.cqltypes
+    orig_serialize = cassandra.cqltypes.UTF8Type.serialize
+    def myserialize(ustr, protocol_version):
+        return ustr.encode('utf-8', errors='surrogateescape')
+    cassandra.cqltypes.UTF8Type.serialize = myserialize
+    try:
+        stmt = cql.prepare(f'INSERT INTO {table1} (k, t) VALUES (1, ?)')
+        for b in good_utf8:
+            print(b)
+            cql.execute(stmt, [b.decode()])
+            results = list(cql.execute(f"SELECT k, t FROM {table1} WHERE k=1"))
+            assert len(results) == 1
+            assert results[0].k == 1 and results[0].t.encode('utf-8') == b
+        for b in bad_utf8:
+            print(b)
+            # Scylla prints "Exception while binding column t: marshaling error:
+            # Validation failed - non-UTF8 character in a UTF8 string, at byte
+            # offset 0". Cassandra prints "String didn't validate.". The only
+            # thing in common is the word 'validat' :-)
+            with pytest.raises(InvalidRequest, match=re.compile('validat', re.IGNORECASE)):
+                cql.execute(stmt, [b.decode(errors='surrogateescape')])
+    finally:
+        cassandra.cqltypes.UTF8Type.serialize = orig_serialize
+
+# 3. We can also insert the bad UTF-8 as part of the request string itself.
+#    This will make the entire request string invalid UTF-8, not just the
+#    value to be inserted, so Scylla and Cassandra should complain that the
+#    entire request is bad - not just the inserted value.
+# FIXME: this test is INCOMPLETE! It's very hard to get Python to output an
+# illegal UTF-8 string in this case, and I gave up. Unlike the prepared-
+# statement case above where it was easy to find and monkey-patch the
+# function responsible for converting the string to bytes, in this case it
+# was harder to find this function and I gave up.
+def test_validation_utf8_query(cql, table1):
+    for b in good_utf8:
+        s = b.decode('utf-8')
+        print(s)
+        cql.execute(f"INSERT INTO {table1} (k, t) VALUES (1, '{s}')")
+        results = list(cql.execute(f"SELECT k, t FROM {table1} WHERE k=1"))
+        assert len(results) == 1
+        assert results[0].k == 1 and results[0].t == s
+    # FIXME: Need to figure out the appropriate monkey-patching or other
+    # trick to make the following work (i.e., pass the invalid string to
+    # the server, and let the server - not the driver - fail.
+    # for b in bad_utf8:
+    #  print(b)
+    #  cql.execute("INSERT INTO {} (k, t) VALUES (1, '{}')".format(table1, b.decode(errors='surrogateescape')))
```

Comment:
Possible trick: hijack the connection between server and client using a proxy, send values such as `XXX replace me XXX`, and replace them in the proxy.

Reply:
Yes, this is possible, and the proxy can even be in Python (the proxy can use bytes instead of unicode strings) so it can be part of the test to run this proxy and connect to it. Python even has an existing way to encode invalid UTF-8 - that "surrogateescape" thing - so we don't necessarily need to invent our own "XXX" things.
Sounds like a very interesting exercise :-)